src/System.Private.ServiceModel/tools/test/SelfHostWcfService/WSRequestReplyService.cs

```diff
@@ -0,0 +1,153 @@
+                log.Add(string.Format("UploadData received {0}", data));
+            }
+            else
+            {
+                log.Add(string.Format("UploadData received {0} length string.", data.Length));
+            }
+
+            // Access the RemoteEndpointMessageProperty
+            RemoteEndpointMessageProperty remp = OperationContext.Current.IncomingMessageProperties[RemoteEndpointMessageProperty.Name] as RemoteEndpointMessageProperty;
+            IPHostEntry hostEntry = Dns.GetHostEntry("PartialTrustEnvironment.GetMachineName()");
+            bool success = false;
+            foreach (IPAddress address in hostEntry.AddressList)
+            {
+                if (remp.Address == address.ToString())
+                {
+                    success = true;
+                    break;
+                }
+            }
+
+            if (!success)
+            {
+                log.Add(RemoteEndpointMessagePropertyFailure);
+            }
+        }
+
+        public string DownloadData()
+        {
+            string data = CreateInterestingString(rand.Next(512, 4096));
+            log.Add(string.Format("DownloadData returning {0} length string.", data.Length));
+            return data;
+        }
+
+        public void UploadStream(Stream stream)
+        {
+            int readResult;
+            int bytesRead = 0;
+            byte[] buffer = new byte[1000];
+            do
+            {
+                readResult = stream.Read(buffer, 0, buffer.Length);
+                bytesRead += readResult;
+            }
+            while (readResult != 0);
+
+            stream.Close();
+
+            log.Add(string.Format("UploadStream read {0} bytes from the client's stream.", bytesRead));
+        }
+
+        public Stream DownloadStream()
+        {
+            log.Add("DownloadStream");
+            localStream = new FlowControlledStream();
+            localStream.ReadThrottle = TimeSpan.FromMilliseconds(500);
+            localStream.StreamDuration = TimeSpan.FromSeconds(1);
+
+            return localStream;
+        }
+
+        public Stream DownloadCustomizedStream(TimeSpan readThrottle, TimeSpan streamDuration)
+        {
+            log.Add("DownloadStream");
+            localStream = new FlowControlledStream();
+            localStream.ReadThrottle = readThrottle;
+            localStream.StreamDuration = streamDuration;
+
+            return localStream;
+        }
+
+        public void ThrowingOperation(Exception exceptionToThrow)
+        {
+            log.Add("ThrowingOperation");
+            throw exceptionToThrow;
+        }
+
+        public string DelayOperation(TimeSpan delay)
+        {
+            log.Add("DelayOperation");
+            Thread.CurrentThread.Join(delay);
+            return "Done with delay.";
+        }
+
+        public List<string> GetLog()
+        {
+            return log;
+        }
+
+        public static string CreateInterestingString(int length)
+        {
```

Comment:
I see some of these constants and methods appearing in more than one place.  Can we move them to some common helper classes?